File: scripts/convert-ngsl-nawl.py
#!/usr/bin/env python3
# Convert NGSL.csv → daily.json, NAWL.csv → toefl.json
import csv, json, os, re

BASE = os.path.join(os.path.dirname(__file__), '..')

def clean(s):
    return re.sub(r'\s+', ' ', s.replace('\xa0', ' ')).strip()

# ── Parse NAWL.csv ────────────────────────────────────────────────────────────
# Columns: Rank, Lemma, POS, Definition, JP Translation
nawl_words = []
nawl_seen  = set()
with open(os.path.join(BASE, 'NAWL.csv'), encoding='utf-8-sig') as f:
    for row in csv.reader(f):
        if not row or row[0].strip() == 'Rank':
            continue
        rank = int(row[0]) if row[0].isdigit() else 9999
        word = clean(row[1]) if len(row) > 1 else ''
        defn = clean(row[3]) if len(row) > 3 else ''
        jp   = clean(row[4]) if len(row) > 4 else ''
        if not word or word.lower() in nawl_seen:
            continue
        nawl_seen.add(word.lower())
        nawl_words.append({'word': word, 'rank': rank, 'definition': defn, 'translation': jp})

nawl_words.sort(key=lambda w: w['rank'])

total  = len(nawl_words)
third  = total // 3
for i, w in enumerate(nawl_words):
    w['level'] = 1 if i < third else (2 if i < third * 2 else 3)

# Build rank map for NGSL cross-reference
rank_map = {w['word'].lower(): w['rank'] for w in nawl_words}

l1 = sum(1 for w in nawl_words if w['level'] == 1)
l2 = sum(1 for w in nawl_words if w['level'] == 2)
l3 = sum(1 for w in nawl_words if w['level'] == 3)
print(f'NAWL: {total} words | ★={l1} ★★={l2} ★★★={l3}')
print(f'  L1 sample: {[w["word"] for w in nawl_words if w["level"]==1][:6]}')
print(f'  L2 sample: {[w["word"] for w in nawl_words if w["level"]==2][:6]}')
print(f'  L3 sample: {[w["word"] for w in nawl_words if w["level"]==3][:6]}')
print(f'  no def: {sum(1 for w in nawl_words if not w["definition"])}')
print(f'  no JP:  {sum(1 for w in nawl_words if not w["translation"])}')

toefl_out = {
    'category': 'toefl',
    'source': 'NAWL with Japanese translations',
    'words': [
        {
            'id': f'f{i+1:04d}',
            'word': w['word'],
            'translation': w['translation'],
            'definition': w['definition'],
            'level': w['level'],
        }
        for i, w in enumerate(nawl_words)
    ]
}
with open(os.path.join(BASE, 'data', 'toefl.json'), 'w', encoding='utf-8') as f:
    json.dump(toefl_out, f, ensure_ascii=False, indent=2)
print(f'\n✅ toefl.json saved ({total} words)')

# ── Parse NGSL.csv ────────────────────────────────────────────────────────────
# Columns: Meanings (word), English Definition, POS, J Translation
ngsl_words = []
ngsl_seen  = set()
with open(os.path.join(BASE, 'NGSL.csv'), encoding='utf-8-sig') as f:
    for row in csv.reader(f):
        if not row or row[0].strip() in ('Meanings', ''):
            continue
        word = clean(row[0]) if len(row) > 0 else ''
        defn = clean(row[1]) if len(row) > 1 else ''
        jp   = clean(row[3]) if len(row) > 3 else ''
        if not word or not re.search(r'[a-zA-Z]', word):
            continue
        if word.lower() in ngsl_seen:
            continue
        ngsl_seen.add(word.lower())

        # Level via NAWL rank cross-reference
        r = rank_map.get(word.lower())
        if r is not None:
            level = 1 if r <= third else (2 if r <= third * 2 else 3)
        else:
            level = 3

        ngsl_words.append({'word': word, 'definition': defn, 'translation': jp, 'level': level})

# Sort: level asc, then alphabetically within level
ngsl_words.sort(key=lambda w: (w['level'], w['word'].lower()))

g_total = len(ngsl_words)
g1 = sum(1 for w in ngsl_words if w['level'] == 1)
g2 = sum(1 for w in ngsl_words if w['level'] == 2)
g3 = sum(1 for w in ngsl_words if w['level'] == 3)
print(f'\nNGSL: {g_total} words | ★={g1} ★★={g2} ★★★={g3}')
print(f'  L1 sample: {[w["word"] for w in ngsl_words if w["level"]==1][:8]}')
print(f'  L2 sample: {[w["word"] for w in ngsl_words if w["level"]==2][:8]}')
print(f'  L3 sample: {[w["word"] for w in ngsl_words if w["level"]==3][:8]}')
print(f'  no def: {sum(1 for w in ngsl_words if not w["definition"])}')
print(f'  no JP:  {sum(1 for w in ngsl_words if not w["translation"])}')

daily_out = {
    'category': 'daily',
    'source': 'NGSL (New General Service List) with Japanese translations',
    'words': [
        {
            'id': f'd{i+1:04d}',
            'word': w['word'],
            'translation': w['translation'],
            'definition': w['definition'],
            'level': w['level'],
        }
        for i, w in enumerate(ngsl_words)
    ]
}
with open(os.path.join(BASE, 'data', 'daily.json'), 'w', encoding='utf-8') as f:
    json.dump(daily_out, f, ensure_ascii=False, indent=2)
print(f'\n✅ daily.json saved ({g_total} words)')
print('\n=== DONE ===')

File: scripts/convert-ngsl-nawl.js
#!/usr/bin/env node
// Convert NGSL.csv → daily.json, NAWL.csv → toefl.json
// NAWL: rank-ordered (rank 1 = most frequent = easiest)
// NGSL: alphabetical → levels assigned via NAWL rank cross-reference

const fs   = require('fs');
const path = require('path');

function parseCSV(text) {
  const lines = text.split('\n');
  const rows  = [];
  for (const line of lines) {
    if (!line.trim()) continue;
    // simple CSV parse: handle quoted fields
    const cols = [];
    let cur = '', inQ = false;
    for (let i = 0; i < line.length; i++) {
      const c = line[i];
      if (c === '"') { inQ = !inQ; }
      else if (c === ',' && !inQ) { cols.push(cur.trim()); cur = ''; }
      else { cur += c; }
    }
    cols.push(cur.trim());
    rows.push(cols);
  }
  return rows;
}

function cleanDef(s) {
  return s.replace(/\xa0/g, ' ').replace(/\s+/g, ' ').trim();
}

// ── Parse NAWL.csv ────────────────────────────────────────────────────────────
// Columns: Rank, Lemma, POS, Definition, JP Translation
const nawlText = fs.readFileSync(path.join(__dirname, '..', 'NAWL.csv'), 'utf8')
  .replace(/^﻿/, ''); // strip BOM
const nawlRows = parseCSV(nawlText).slice(1); // skip header

// Deduplicate: keep first occurrence of each word
const nawlSeen = new Set();
const nawlWords = [];
for (const row of nawlRows) {
  const word = (row[1] || '').trim();
  const rank = parseInt(row[0]) || 9999;
  const def  = cleanDef(row[3] || '');
  const jp   = (row[4] || '').trim();
  if (!word || nawlSeen.has(word.toLowerCase())) continue;
  nawlSeen.add(word.toLowerCase());
  nawlWords.push({ word, rank, definition: def, translation: jp });
}

nawlWords.sort((a, b) => a.rank - b.rank); // ensure sorted by rank

const nawlTotal = nawlWords.length;
const nawlThird = Math.floor(nawlTotal / 3);
nawlWords.forEach((w, i) => {
  w.level = i < nawlThird ? 1 : i < nawlThird * 2 ? 2 : 3;
});

// Build rank map for NGSL level cross-reference
const nawlRankMap = new Map();
nawlWords.forEach(w => nawlRankMap.set(w.word.toLowerCase(), w.rank));

console.log(`NAWL: ${nawlTotal} unique words`);
const n1=nawlWords.filter(w=>w.level===1).length;
const n2=nawlWords.filter(w=>w.level===2).length;
const n3=nawlWords.filter(w=>w.level===3).length;
console.log(`  Level distribution: ★=${n1} ★★=${n2} ★★★=${n3}`);
console.log(`  Sample L1: ${nawlWords.filter(w=>w.level===1).slice(0,6).map(w=>w.word).join(', ')}`);
console.log(`  Sample L3: ${nawlWords.filter(w=>w.level===3).slice(-6).map(w=>w.word).join(', ')}`);

// ── Save toefl.json ───────────────────────────────────────────────────────────
const toeflOut = {
  category: 'toefl',
  source: 'NAWL (New Academic Word List) with Japanese translations',
  words: nawlWords.map((w, i) => ({
    id: `f${String(i+1).padStart(4,'0')}`,
    word: w.word,
    translation: w.translation,
    definition: w.definition,
    level: w.level,
  }))
};
fs.writeFileSync(path.join(__dirname, '..', 'data', 'toefl.json'),
  JSON.stringify(toeflOut, null, 2));
console.log(`\n✅ toefl.json saved (${nawlTotal} words)`);

// ── Parse NGSL.csv ────────────────────────────────────────────────────────────
// Columns: Meanings (word), English Definition, POS, J Translation
const ngslText = fs.readFileSync(path.join(__dirname, '..', 'NGSL.csv'), 'utf8')
  .replace(/^﻿/, ''); // strip BOM
const ngslRows = parseCSV(ngslText).slice(1); // skip header

const ngslSeen = new Set();
const ngslWords = [];
for (const row of ngslRows) {
  const word = (row[0] || '').trim();
  const def  = cleanDef(row[1] || '');
  const jp   = (row[3] || '').trim();
  if (!word || ngslSeen.has(word.toLowerCase())) continue;
  ngslSeen.add(word.toLowerCase());

  // Assign level via NAWL rank cross-reference
  const rank = nawlRankMap.get(word.toLowerCase());
  let level;
  if (rank !== undefined) {
    level = rank <= nawlThird ? 1 : rank <= nawlThird * 2 ? 2 : 3;
  } else {
    level = 3; // not in NAWL (rare word)
  }

  ngslWords.push({ word, definition: def, translation: jp, level });
}

// Sort by level then alphabetically within level
ngslWords.sort((a, b) => a.level - b.level || a.word.localeCompare(b.word));

const ngslTotal = ngslWords.length;
const g1=ngslWords.filter(w=>w.level===1).length;
const g2=ngslWords.filter(w=>w.level===2).length;
const g3=ngslWords.filter(w=>w.level===3).length;
console.log(`\nNGSL: ${ngslTotal} unique words`);
console.log(`  Level distribution: ★=${g1} ★★=${g2} ★★★=${g3}`);
console.log(`  Sample L1: ${ngslWords.filter(w=>w.level===1).slice(0,6).map(w=>w.word).join(', ')}`);
console.log(`  Sample L3: ${ngslWords.filter(w=>w.level===3).slice(0,6).map(w=>w.word).join(', ')}`);

// ── Save daily.json ───────────────────────────────────────────────────────────
const dailyOut = {
  category: 'daily',
  source: 'NGSL (New General Service List) with Japanese translations',
  words: ngslWords.map((w, i) => ({
    id: `d${String(i+1).padStart(4,'0')}`,
    word: w.word,
    translation: w.translation,
    definition: w.definition,
    level: w.level,
  }))
};
fs.writeFileSync(path.join(__dirname, '..', 'data', 'daily.json'),
  JSON.stringify(dailyOut, null, 2));
console.log(`\n✅ daily.json saved (${ngslTotal} words)`);

// ── Missing check ─────────────────────────────────────────────────────────────
console.log('\n── Missing data check ──');
const tNoDef = nawlWords.filter(w=>!w.definition).length;
const tNoJp  = nawlWords.filter(w=>!w.translation).length;
const gNoDef = ngslWords.filter(w=>!w.definition).length;
const gNoJp  = ngslWords.filter(w=>!w.translation).length;
console.log(`TOEFL (NAWL): no def=${tNoDef}, no JP=${tNoJp}`);
console.log(`Daily (NGSL): no def=${gNoDef}, no JP=${gNoJp}`);
console.log('\n=== DONE ===');
